app/controllers/bankdetails/BankAccountNonUKController.scala

```diff
@@ -0,0 +1,89 @@
+ */
+
+package controllers.bankdetails
+
+import cats.data.OptionT
+import cats.implicits._
+import connectors.DataCacheConnector
+import forms.{EmptyForm, Form2, InvalidForm, ValidForm}
+import javax.inject.{Inject, Singleton}
+import models.bankdetails.Account._
+import models.bankdetails.{BankAccount, BankDetails, NonUKAccountNumber}
+import services.StatusService
+import uk.gov.hmrc.play.audit.http.connector.AuditConnector
+import utils.{AuthAction, StatusConstants}
+
+import scala.concurrent.Future
+
+@Singleton
+class BankAccountNonUKController @Inject()(val dataCacheConnector: DataCacheConnector,
+                                           val authAction: AuthAction,
+                                           val auditConnector: AuditConnector,
+                                           val statusService: StatusService) extends BankDetailsController {
+
+  def get(index: Int, edit: Boolean = false) = authAction.async{
+      implicit request =>
+        for {
+          bankDetails <- getData[BankDetails](request.credId, index)
+          status <- statusService.getStatus(request.amlsRefNumber, request.accountTypeId, request.credId)
+        } yield bankDetails match {
+          case Some(x@BankDetails(_, _, Some(BankAccount(_, _, Some(data@NonUKAccountNumber(_)))), _, _, _, _)) if x.canEdit(status) =>
+            Ok(views.html.bankdetails.bank_account_account_number_non_uk(Form2[NonUKAccountNumber](data), edit, index))
+          case Some(x) if x.canEdit(status) =>
+            Ok(views.html.bankdetails.bank_account_account_number_non_uk(EmptyForm, edit, index))
+          case _ => NotFound(notFoundView)
+        }
+  }
+
+  def post(index: Int, edit: Boolean = false) = authAction.async {
+      implicit request => {
+
+        lazy val sendAudit = for {
+          details <- OptionT(getData[BankDetails](request.credId, index))
+          result <- OptionT.liftF(auditConnector.sendEvent(audit.AddBankAccountEvent(details)))
+        } yield result
+
+        Form2[NonUKAccountNumber](request.body) match {
+          case f: InvalidForm =>
+            Future.successful(BadRequest(views.html.bankdetails.bank_account_account_number_non_uk(f, edit, index)))
+          case ValidForm(_, data) =>
+            updateDataStrict[BankDetails](request.credId, index) { bd =>
+              bd.copy(
+                bankAccount = Option(bd.bankAccount.getOrElse(BankAccount(None, None, None)).account(data)),
+                status = Some(if (edit) {
+                  StatusConstants.Updated
+                } else {
+                  StatusConstants.Added
+                })
+              )
+            }.flatMap { _ =>
+              if (edit) {
+                Future.successful(Redirect(routes.SummaryController.get(index)))
+              } else {
+                lazy val redirect = Redirect(routes.SummaryController.get(index))
+                (sendAudit map { _ =>
+                  redirect
+                }) getOrElse redirect
```

Comment:
Possible refactor here to make this easier to follow? Perhaps split out the above into methods that do a single purpose? With particular reference to the update on the object.